solcore/absorption_calculator/rigorous_coupled_wave.py

```diff
@@ -246,6 +307,118 @@ def rcwa_absorption_per_layer(S, n_layers):
     for i1, layer in enumerate(np.arange(n_layers-2)+2):
         A[i1] = np.real(sum(S.GetPowerFlux('layer_' + str(layer))) - sum(S.GetPowerFlux('layer_' + str(layer+1))))
 
-    A = [x if x > 0 else 0 for x in A]
+    A = np.array([x if x > 0 else 0 for x in A])
+
+    return A
+
+def RCWA_wl(wl, geom_list, l_oc, s_oc, s_names, pol, theta, phi, widths, size, orders, rcwa_options):
```

Comment:
In this function there's a lot of code repetition. The following is executed esentially the same in 4 different places:

```python
S.SetExcitationPlanewave((theta, phi), s, p, 0)
S.SetFrequency(1 / wl)
out = rcwa_rat(S, len(widths))
R = out['R']
T = out['T']
A_layer = rcwa_absorption_per_layer(S, len(widths))
```

I think this should be a function, possibly internal to `RCWA_wl` defined just before starting the `if...else` for the polarizations, looking something like:

```python
def vs_pol(s, p):
    S.SetExcitationPlanewave((theta, phi), s, p, 0)
    S.SetFrequency(1 / wl)
    out = rcwa_rat(S, len(widths))
    R = out['R']
    T = out['T']
    A_layer = rcwa_absorption_per_layer(S, len(widths))
    return R, T, A_layer
```

And then call it as:

```python
if len(pol) == 2:
    R, T, A_layer = vs_pol(pol[0], pol[1])
elif pol in "sp":
    R, T, A_layer = vs_pol(int(pol == "s"), int(pol == "p"))
else:
    R_s, T_s, A_layer_s = vs_pol(1, 0)
    R_p, T_p, A_layer_p = vs_pol(0, 1)
    R = (R_s + R_p) / 2
    T = (R_s + R_p) / 2
    A_layer = (A_layer_s + A_layer_p) / 2
```
Unless I'm missing something, this is more concise, I believe.

Reply:
Yes, much better, and a similar change can also be made in the profile calculation!